test/e2e/network/service.go

```diff
@@ -2947,6 +2951,179 @@ var _ = SIGDescribe("Services", func() {
 	})
+
+	ginkgo.It("should complete a service status lifecycle", func() {
+
+		cs := f.ClientSet
+		dc := f.DynamicClient
+		ns := f.Namespace.Name
+		svcResource := schema.GroupVersionResource{Group: "", Version: "v1", Resource: "services"}
+		testSvcName := "test-service-" + utilrand.String(5)
+		testSvcLabels := map[string]string{"test-service-static": "true"}
+		testSvcLabelsFlat := "test-service-static=true"
+
+		w := &cache.ListWatch{
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				options.LabelSelector = testSvcLabelsFlat
+				return cs.CoreV1().Services(ns).Watch(context.TODO(), options)
+			},
+		}
+
+		svcList, err := cs.CoreV1().Services("").List(context.TODO(), metav1.ListOptions{LabelSelector: testSvcLabelsFlat})
+		framework.ExpectNoError(err, "failed to list Pods")
+
+		ginkgo.By("creating a Service")
+		testService := v1.Service{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:   testSvcName,
+				Labels: testSvcLabels,
+			},
+			Spec: v1.ServiceSpec{
+				Type: "ClusterIP",
+				Ports: []v1.ServicePort{{
+					Name:       "http",
+					Protocol:   v1.ProtocolTCP,
+					Port:       int32(80),
+					TargetPort: intstr.FromInt(80),
+				}},
+			},
+		}
+		_, err = cs.CoreV1().Services(ns).Create(context.TODO(), &testService, metav1.CreateOptions{})
+		framework.ExpectNoError(err, "failed to create Service")
+
+		ginkgo.By("watching for the Service to be added")
+		ctx, cancel := context.WithTimeout(context.Background(), svcReadyTimeout)
+		defer cancel()
+		_, err = watchtools.Until(ctx, svcList.ResourceVersion, w, func(event watch.Event) (bool, error) {
+			if svc, ok := event.Object.(*v1.Service); ok {
+				found := svc.ObjectMeta.Name == testService.ObjectMeta.Name &&
+					svc.ObjectMeta.Namespace == ns &&
+					svc.Labels["test-service-static"] == "true"
+				if !found {
+					framework.Logf("observed Service %v in namespace %v with labels: %v & ports %v", svc.ObjectMeta.Name, svc.ObjectMeta.Namespace, svc.Labels, svc.Spec.Ports)
+					return false, nil
+				}
+				framework.Logf("Found Service %v in namespace %v with labels: %v & ports %v", svc.ObjectMeta.Name, svc.ObjectMeta.Namespace, svc.Labels, svc.Spec.Ports)
+				return found, nil
+			}
+			framework.Logf("Observed event: %+v", event.Object)
+			return false, nil
+		})
+		framework.ExpectNoError(err, "failed to locate Service %v in namespace %v", testService.ObjectMeta.Name, ns)
+		framework.Logf("Service %s created", testSvcName)
+
+		ginkgo.By("patching the ServiceStatus")
+		serviceStatusPatch, err := json.Marshal(map[string]interface{}{
+			"metadata": map[string]interface{}{
+				"labels": map[string]string{"test-service": "patched"},
+			},
+			"spec": map[string]interface{}{
+				"ports": []map[string]interface{}{{
+					"name":       "http8080",
+					"port":       int32(8080),
+					"targetPort": int(8080),
+					"selector": []map[string]interface{}{{
+						"type": "LoadBalancer",
+					}},
+				}},
+			},
+		})
```

Comment:
don´t you have to patch the Status too?
https://kubernetes.io/docs/reference/kubernetes-api/services-resources/service-v1/#ServiceStatus

```
status:
  loadBalancer:
    ingress:
    - ip: 203.0.113.1
  ```

Reply:
> don´t you have to patch the Status too?
> https://kubernetes.io/docs/reference/kubernetes-api/services-resources/service-v1/#ServiceStatus
> 
> ```
> status:
>   loadBalancer:
>     ingress:
>     - ip: 203.0.113.1
> ```

When tracking this test with apisnoop the test does hit the `patchCoreV1NamespacedServiceStatus` endpoint. The test approach was reviewed with @smarterclayton in the [Jan 12 conformance meeting](https://github.com/kubernetes/kubernetes/issues/94867#issuecomment-759016564).

Reply:
that's confusing to me as well. do we have an explanation why this test hits status? I mean, we list the status sub-resource, but we don't list any changes to it. @smarterclayton is that sufficient to say it works?

Reply:
I asked Clayton in that [conformance meeting](https://www.youtube.com/watch?v=ENXq4UyQr9I&t=488) if the test process was on the right track:
>  If you are patching status you may mutate the both status and metadata even though it it is s a status endpoint there are historical reasons for that and we couldn't redefine it. And in the spec the  full object should be returned, but only mutations in status and metadata should be accepted. @smarterclayton 

See also mailing list message from @hh https://groups.google.com/g/kubernetes-sig-architecture/c/Ekv2NF2D5B8

Reply:
But that says: "you may mutate both status and metadata", but the problem is we are *not* mutating status. I think we should be.

Reply:
This is the audit log entry from [pull-kubernetes-e2e-gce-ubuntu-containerd #1352299357109489664](https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/98018/pull-kubernetes-e2e-gce-ubuntu-containerd/1352299357109489664/) that shows the patching of the endpoint.

```
curl -L \
https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/98018/pull-kubernetes-e2e-gce-ubuntu-containerd/1352299357109489664/artifacts/e2e-5d7b0f355d-a7d53-master/kube-apiserver-audit.log-20210121-1611249901.gz  \
| gzip -d   \
| grep "should complete a service status lifecycle" \
| grep patch \
| jq . 
```

```json
{
"requestURI": "/api/v1/namespaces/services-8164/services/test-service-szbkj/status",
  "verb": "patch",
 "userAgent": "e2e.test/v1.21.0 (linux/amd64) kubernetes/4405248 -- \
[sig-network] Services should complete a service status lifecycle",
   "responseStatus": {
    "metadata": {},
    "code": 200
  },
 "requestObject": {
    "metadata": {
      "labels": {
        "test-service": "patched"
      }
    },
    "spec": {
      "ports": [
        {
          "name": "http8080",
          "port": 8080,
          "selector": [
            {
              "type": "LoadBalancer"
            }
          ],
          "targetPort": 8080
        }
      ]
    }
  }
}
```
```

Reply:
Sure, but it doesn't actually *mutate* the status. It just mutates the metadata. So what we're really testing is "can I mutate metadata via the status endpoint" which is nice and all but it's not the primary purpose of the status endpoint.

Is there any field in the status we can mutate that won't cause a race with the controller?

Reply:
After a brief convo with @liggitt I realized that since we now use watch, we shouldn't have an issue with races with the controller, since we will see our change (even if we do see the controller change it back *after* that, we can ignore it).

so, you should be patching a specific field in status. just pick one...

Reply:
What if we create an arbitrary `status.condition` ?
```
# kubectl explain service.status.conditions
KIND:     Service
VERSION:  v1

RESOURCE: conditions <[]Object>

DESCRIPTION:
     Current service state

     Condition contains details for one aspect of the current state of this API
     Resource.

FIELDS:
   lastTransitionTime	<string> -required-
     lastTransitionTime is the last time the condition transitioned from one
     status to another. This should be when the underlying condition changed. If
     that is not known, then using the time when the API field changed is
     acceptable.

   message	<string> -required-
     message is a human readable message indicating details about the
     transition. This may be an empty string.

   observedGeneration	<integer>
     observedGeneration represents the .metadata.generation that the condition
     was set based upon. For instance, if .metadata.generation is currently 12,
     but the .status.conditions[x].observedGeneration is 9, the condition is out
     of date with respect to the current state of the instance.

   reason	<string> -required-
     reason contains a programmatic identifier indicating the reason for the
     condition's last transition. Producers of specific condition types may
     define expected values and meanings for this field, and whether the values
     are considered a guaranteed API. The value should be a CamelCase string.
     This field may not be empty.

   status	<string> -required-
     status of the condition, one of True, False, Unknown.

   type	<string> -required-
     type of condition in CamelCase or in foo.example.com/CamelCase.
```

Reply:
this is the more realistic , I guess
```
status:
  loadBalancer:
    ingress:
    - ip: 203.0.113.1
```

Reply:
Looks like this is taken care of as well.

```
+		ginkgo.By("patching the ServiceStatus")
+		lbStatus := v1.LoadBalancerStatus{
+			Ingress: []v1.LoadBalancerIngress{{IP: "203.0.113.1"}},
+		}
```